crates/cxx-qt-build/src/dir.rs

```diff
@@ -81,3 +81,75 @@ pub(crate) fn out() -> PathBuf {
 pub(crate) fn is_exporting() -> bool {
     export().is_some()
 }
+
+#[cfg(unix)]
+pub(crate) fn symlink_or_copy_directory(
+    source: impl AsRef<Path>,
+    dest: impl AsRef<Path>,
+) -> Result<bool> {
+    match std::os::unix::fs::symlink(&source, &dest) {
+        Ok(()) => Ok(true),
+        Err(e) if e.kind() != std::io::ErrorKind::AlreadyExists => Err(e),
+        // Two dependencies may be reexporting the same shared dependency, which will
+        // result in conflicting symlinks.
+        // Try detecting this by resolving the symlinks and checking whether this leads us
+        // to the same paths. If so, it's the same include path for the same prefix, which
+        // is fine.
+        Err(_) => Ok(fs::canonicalize(source)? == fs::canonicalize(dest)?),
+    }
+}
+
+#[cfg(not(unix))]
+pub(crate) fn symlink_or_copy_directory(
+    source: impl AsRef<Path>,
+    dest: impl AsRef<Path>,
+) -> Result<bool> {
+    deep_copy_directory(source.as_ref(), dest.as_ref())
+}
+
+#[cfg(not(unix))]
+fn deep_copy_directory(source: &Path, dest: &Path) -> Result<bool> {
+    fs::create_dir_all(dest)?;
+    for entry in fs::read_dir(source)? {
+        let entry = entry?;
+        let source_path = entry.path();
+        let dest_path = dest.join(entry.file_name());
+        if entry.file_type()?.is_dir() {
+            if deep_copy_directory(&source_path, &dest_path)? {
+                continue;
+            }
+            return Ok(false);
+        }
+        if !dest_path.try_exists()? {
+            fs::copy(&source_path, &dest_path)?;
+            continue;
+        }
+        if files_conflict(&source_path, &dest_path)? {
+            return Ok(false);
+        }
+    }
+    Ok(true)
+}
+
+#[cfg(not(unix))]
+fn files_conflict(source: &Path, dest: &Path) -> Result<bool> {
+    use fs::File;
+    use std::io::Read;
+    let mut source = File::open(source)?;
+    let mut dest = File::open(dest)?;
+    if source.metadata()?.len() != dest.metadata()?.len() {
+        return Ok(true);
+    }
+    let mut source_buf = [0; 1024];
+    let mut dest_buf = [0; 1024];
+    loop {
+        let source_n = source.read(&mut source_buf)?;
+        let dest_n = dest.read(&mut dest_buf)?;
+        if source_n == 0 && dest_n == 0 {
+            return Ok(false);
+        }
+        if &source_buf[..source_n] != &dest_buf[..dest_n] {
+            return Ok(true);
+        }
+    }
```

Comment:
This code feels quite C-esque, with the indexing and manual buffer management.

I prefer to see Rusts `ReadBuffer` and `Bytes` iterator here, which should do pretty much the same but express the intent more clearly.
See also the example code here: https://users.rust-lang.org/t/efficient-way-of-checking-if-two-files-have-the-same-content/74735

Please refactor this to simply iterate and compare the zipped bytes.
The check for metadata path is still useful to keep though, as that offers a fast path to failure.

Reply:
Hmm. I agree with this comment on that thread:

>I'm not sure if `.bytes()` iterator is the best for it, since handling of `io::Error` adds overhead and can't be optimized out. Reading in larger chunks should improve it.

What would you think about using the `BufReader` buffers themselves for comparison, as in [the latest commit](https://github.com/KDAB/cxx-qt/pull/1136/files#diff-b3d0f156e4bfb7954ff4537ae918929c3edfe9356658c3a8d7b99bb2f360ee2f)? Asking Rust to do an equality check on two slices definitely seems more efficient than equality-checking byte by byte, unwrapping a `Result` every time.

Reply:
Hm, this actually seems like a good compromise between readability and indexing complexity.

If you can modify the method this way I can approve the PR :)

Reply:
Done! Excited to see this in.